Issue 22574 - Compiler segfaults on `template test(x* x)`

maxhaton opened the pull request.

maxhaton committed:
Fix Issue 22574 - Compiler segfaults on `template test(x* x)`

thewilsonator reviewed (approved):
If this fixes issue 22574 then please make the commit message say "Fix issue 22574: ..."

RazvanN7 commented on test/fail_compilation/test22574.d:
```diff
@@ -0,0 +1,13 @@
+TEST_OUTPUT:
+---
+fail_compilation/test22574.d(100): Error: variable `x` is used as a type
+fail_compilation/test22574.d(100):        variable `x` is declared here
```
Can we agree that this is a weird error message? When looking at `x* x` the compiler should first try to analyze the type and then the variable name, however, in this case it seems that the name is inserted in the symbol table before actually resolving the type. I would expect that the error message is something along the lines of "type `x` is not defined".

RazvanN7 replied:
Looking in [template declaration semantic](https://github.com/dlang/dmd/blob/master/src/dmd/dsymbolsem.d#L2560) you can see that when the parameter is [declared](https://github.com/dlang/dmd/blob/master/src/dmd/dtemplate.d#L5534) (for template value parameters) the identifier is inserted in the scope before actually doing any semantic on the type. However, the next thing that it is done in [template declaration semantic](https://github.com/dlang/dmd/blob/master/src/dmd/dsymbolsem.d#L2566) is to do [semantic on the type](https://github.com/dlang/dmd/blob/master/src/dmd/templateparamsem.d#L78). So, if we want to fix the underlying issue and report a correct error, we should add

```d
if (valtype)
    valtype = valtype.typeSemantic(loc, sc);
```

in TemplateValueParameter::declareParameter.

That will make the segfault go away and output: "Undefined identifier `x`"

maxhaton replied:
This is why I left the comment in the source code.

RazvanN7 reviewed (changes requested):
See my previous comments

ibara committed:
Fix build for Solaris/Illumos
Fix Issue 22573 - DMD compiler errors on Illumos/Solaris

maxhaton committed:
Incorporate razvan review

RazvanN7 reviewed (approved):
Thanks!

maxhaton commented:
@RazvanN7 I have changed it as you suggested, however this has led to some changes in error messages in other (failing, to be clear) template value parameter tests. I have corrected these error messages since I think they're basically reasonable still

RazvanN7 commented:
Please add a commit message that contains "Fix Issue 22574" so that the bot picks it up.
